File: v2/server/js/Problems/Fly.js


var dStay   = 0; 
var dUp     = 1;
var dDown   = 2;
var dLeft   = 3; 
var dRight  = 4; 

var lt = function(x,y){
  return (x<=y) ;
};

var head_ = function( xs ){
  if( xs.length == 0 ) return null;
  return xs[0];
};

var avg = function( poses ){

  if( poses.length == 0 ) return null;

  var sum0 = 0;
  var sum1 = 0;

  for( var i in poses ){
    var pos = poses[i];
    sum0 += pos[0];
    sum1 += pos[1];
  }

  return [ Math.round( sum0 / poses.length ) , Math.round( sum1 / poses.length ) ];
};

var output_ = function( dir_ ){
  return dir_ ;
};

var myApplePoses_ = function( input_ ){
  return input_[0];
};

var nearestFlyPos_ = function( input_ ){
  return input_[1];
};

var myPos_ = function( input_ ){
  return input_[2];
};

var inputEnergy_ = function( input_ ){
  return input_[3];
};


var posToDir_ = function( posMy , posHer ){ 
      
  if( posHer === null ){
    return dStay ;
  }
       
  var d = minus( posHer , posMy );
  var dx = d[0] , dy = d[1];
       
  if(   dx  > dy && (-dx) > dy ) { return dUp   ; }
  if(   dx  > dy               ) { return dRight; }
  if( (-dx) > dy               ) { return dLeft ; }
  if( true                     ) { return dDown ; }
      
};

var minus = function( pos1 , pos2 ){ 
  return [ pos1[0]-pos2[0] , pos1[1]-pos2[1] ]; 
};

var dist = function( pos1 , pos2 ) {
    var d = minus( pos1 , pos2 );
    var dx = d[0] , dy = d[1];
    return Math.sqrt(  dx*dx + dy*dy );
};


$(document).ready(function() {

console.log( "Fly.js ready!" );  





var Fly;
Global.Problems.fly = Fly = {

  progLib : {
    prog1 : function (x0){return output_(posToDir_(  myPos_(x0)   , head_(myApplePoses_(x0))  )   );},
    prog2 : function (x0){return output_(dRight);},
    prog3 : function (x0){return output_(posToDir_(  myPos_(x0)   , nearestFlyPos_(x0) )  );},
  },

//prog_1 input = output_ $ posToDir_ (myPos_ input) (head_ $ myApplePoses_ input)
//prog_2 _     = output_ $ dRight 
//prog_3 input = output_ $ posToDir_ (myPos_ input) (nearestFlyPos_ input)



  onLoadedProblemData : function( data ){
    //console.log( data );
    this.loadResources();
    this.dataJSON = JSON.stringify( data.world );
    this.loadFromDataJSON( );
  },

  restart : function( solution , drawIt ){

    if( solution === undefined ){
      solution = this.defaultSolution ;
      this.actualSolution = solution  ;
    }
    if( drawIt   === undefined ){drawIt   = true;}

    this.loadFromDataJSON( );
    this.mapaAt( this.solutionFlyPos ).prog = solution ;
    if(drawIt){ this.drawMapa(); }
  },

  loadFromDataJSON : function( ){

    var data = JSON.parse( this.dataJSON );

    this.mapa = data.mapa ;
    
    this.mapa[data.solutionFlyPos[0]][data.solutionFlyPos[1]] = {  
      type       : 'fly' ,
      progName   : '_'   ,
      energy     : 1  
    };

    this.solutionFlyPos = data.solutionFlyPos ;

    this.fliesToDo      = data.fliesToDo;
    this.doneFlies      = data.doneFlies;
    this.applePoses     = data.applePoses;
    
    this.numSteps       = data.numSteps;

    this.fliesToDo.unshift( data.solutionFlyPos );

    for( var y in this.mapa ){
      for( var x in this.mapa[y] ){
        var obj = this.mapa[x][y] ;
        if( obj.type === 'fly' ){
          obj.prog = this.progLib[ obj.progName ] ;
        }
      }
    }

  },

  loadResources : function(){

    //console.log( this );

    this.antImg = {};

    var dirs = [ 'u' , 'd' , 'l' , 'r' ];

    for( var i in dirs ){
      this.antImg[ dirs[i] ] = new Image();  
      this.antImg[ dirs[i] ].src = 'img/ant-'+ dirs[i] +'.png';      
    }

    this.flyImg = new Image();
    this.flyImg.src = 'img/fly.png' ;

    this.appleImg = new Image();
    this.appleImg.src = 'img/apple.png' ;

    this.wallImg = new Image();
    this.wallImg.src = 'img/wall.png' ;
  },

  onGo : function(){

    this.solutions     = [] ;
    this.i             = 0  ;

    $('#special').html( 
      '<b>Each number</b> (=fitness) <b>link represents one best-of-generation individual. Click it and see!</b>'+
      '<div id="fly-links"></div><br>' +
      '<table><tr><td><div id="fly-speed">Speed&nbsp;&nbsp;&nbsp;</div>' +
      '<td><div id="fly-speed-slider" style="font-size: 75%;width:600px"></div></table><br>'+
      '<table><tr><td><div id="fly-speed">Time&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</div>' +
      '<td><div id="fly-time-slider" style="font-size: 75%;width:600px"></div></table><br>'+
      '<canvas id="fly-canvas" width="666" height="666"></canvas>' +
      '<div id="fly-info-container" style="font-size:66%"></div>' );


    this.ctx = $('#fly-canvas')[0].getContext('2d');

    //$(function() {
    //  $( "#fly-speed-slider" ).slider();
    //});

    var that = this;

    this.maxSpeed  = 500 ;
    this.initSpeed = 450 ;

    $(function() {
      $( "#fly-speed-slider" ).slider({
        range: "min",
        value: that.initSpeed ,
        min: 0,
        max: that.maxSpeed,
        slide: function( event, ui ) {
          that.changeSpeed( ui.value );
        }
      });
      that.changeSpeed( $( "#fly-speed-slider" ).slider( "value" ) );
    });

    $(function() {
      $( "#fly-time-slider" ).slider({
        range: "min",
        value: 0 ,
        min: 0,
        max: that.numSteps,
        slide: function( event, ui ) {
          $( "#fly-speed-slider" ).slider( "value" , 0 );
          that.changeSpeed( 0 );
          that.goToStepN( ui.value );
        }
      });
      //that.changeSpeed( $( "#fly-time-slider" ).slider( "value" ) );
    });

    this.drawMapa( );

  },

  changeSpeed : function( newSpeed ){
    this.speed = this.maxSpeed - newSpeed;
  },

  fenotyp : function( solution , ffVal ){
    this.solutions[this.i] = solution ;
    $('#fly-links').append( '<a href="#" title="'+ solution +'" '+
      'onclick="Global.Problems.fly.flyLink('+this.i+')">' +ffVal + '</a> ' );

    if( this.i == 0 ){
      //console.log('roflik');
      this.defaultSolution = solution ;
      this.mapaAt( this.solutionFlyPos ).prog = this.defaultSolution ;
    }

    this.i ++ ;

  },

  flyLink : function( i ){
    this.actualSolution = this.solutions[i];
    //console.log('lol : ' +i+ ' ' + this.actualSolution );
    this.restart( this.actualSolution );
    this.run();
  },

  run : function( numSteps ){

    var delta = 1 ;

    if( numSteps === undefined ){
      numSteps = this.numSteps ;
    }

    if( numSteps == 0 ){ 
      return ; 
    } else {
      
      if( ! this.paused ){
        this.step();
        this.drawMapa( ) ;
        $( "#fly-time-slider" ).slider( "value" , this.numSteps - numSteps ) ;
      } else {
        delta = 0 ;
      }
      
      this.paused = this.speed == this.maxSpeed ;
      
      var that = this;
      setTimeout( function(){that.run(numSteps - delta );} , this.speed );  
    }
    
  },

  goToStepN : function( n ){
    this.restart( this.actualSolution , false );
    
    for( var i = 0 ; i < n ; i++  ){
      this.step();
    }

    this.drawMapa();
  },

  step : function(){
    if( this.fliesToDo.length == 0 ){
      this.fliesToDo = this.doneFlies.reverse();
      this.doneFlies = [];
    } else {
      this.stepCurrentFly();
      this.step();
    }
  },

  stepCurrentFly : function(){

    if( this.fliesToDo.length == 0 ){ throw "There should be a fly!" ; }

    var currentFlyPos = this.fliesToDo.shift();
    var fly           = this.mapaAt( currentFlyPos );
    var input         = this.prepareInput( currentFlyPos , fly ); 
    var output        = fly.prog( input ); 

    //console.log( fly.progName + ' ' + output );

    var newFlyPos     = this.tryToMoveFly( output , currentFlyPos ) ;
    
    this.doneFlies.unshift( newFlyPos );

  },



  prepareInput : function( flyPos, fly ){

    return [ this.getSortedPoses( flyPos , this.applePoses ) ,
             this.getNearestFlyPos( flyPos ) ,
             flyPos,
             fly.energy ];
  },


  tryToMoveFly : function( dir , pos ){

    var newPos = this.posPlusDir( pos , dir );

    switch( this.mapaAt(newPos).type ){
      case 'free'  : this.moveFromTo(pos,newPos); return newPos ;  
      case 'apple' : this.eatApple(pos,newPos)  ; return newPos ; 
      default      : /* nothing.. */              return pos    ; 
    }


  },


  moveFromTo : function( from , to ){
    var obj = this.mapaAt( from );
    this.deleteOnPos( from );
    this.putObjOnPos( to , obj )
  },

  // eatApple :: Pos -> Pos -> World -> World
  // eatApple flyPos applePos w0 = 
  //   let w1 = moveFromTo flyPos applePos w0
  //       w2 = updateEnergy (+1) applePos w1
  //    in w2 { applePoses = delete applePos (applePoses w2) }

  eatApple : function( flyPos , applePos ){
    this.moveFromTo( flyPos , applePos );
    this.mapaAt(applePos).energy ++;
    this.deleteFromPoses( this.applePoses , applePos );
    
  },

  deleteOnPos : function( pos ){
    this.mapa[pos[0]][pos[1]] = { type : 'free' };
  },

  deleteFromPoses : function( poses , what ){
    var index = undefined;
    for( var i in poses ){
      if( poses[i][0] == what[0] && poses[i][1] == what[1] ){
        index = i;
        break;
      }
    }

    if( index !== undefined ){
      poses.splice(index, 1);
    }
  },

  putObjOnPos : function( pos , obj ){
    this.mapa[pos[0]][pos[1]] = obj;
  },

  posPlusDir : function( pos , dir ){
    var x = pos[0];
    var y = pos[1];

    switch( dir ){
      case dUp    : return [x  ,y-1] ; 
      case dDown  : return [x  ,y+1] ; 
      case dLeft  : return [x-1,y  ] ; 
      case dRight : return [x+1,y  ] ; 
      case dStay  : return [x  ,y  ] ; 
    }
  },


//getSortedPoses :: World -> Pos -> [Pos] -> [Pos]
//getSortedPoses w pos poses = 
//  map snd . sort . map (\pos'->(dist pos pos',pos')) $ poses

  getSortedPoses : function( pos , poses ){


    return _.chain( poses )
            .map( function( p ){ return [ dist(pos,p) , p ]; }  )
            .sortBy( function(x){return x[0];} )
            .map(  function(x){return x[1];} )
            .value();


  },

  getNearestPos : function( pos , poses ){

    var bestPos  = undefined ;
    var bestDist = 999999999 ;

    for( var i in poses ){
      var actPos = poses[i];
      var dista = dist( pos , actPos );
      if( dista < bestDist  ){
        bestDist = dista;
        bestPos  = actPos;
      }
    }

    if( bestPos === undefined ){ return null ;}
    return bestPos;
  },

  isSameAs : function( pos1 , pos2 ){
    if( pos1 === undefined && pos2 !== undefined ){ return false; }
    if( pos1 !== undefined && pos2 === undefined ){ return false; }

    return ( pos1[0] == pos2[0] && pos1[1] == pos2[1] ) ;
  },

  getNearestApplePos : function( pos ){
    return this.getNearestPos( pos , this.applePoses );
  },

  getNearestFlyPos : function( pos ){
    return this.getNearestPos( pos , this.fliesToDo.concat( this.doneFlies ) );
  },







  mapaAt : function( pos ){
    return this.mapa[pos[0]][pos[1]] ;
  },

  drawMapa : function( ){

    var mapa = this.mapa;

    this.ctx.fillStyle = '#CFE6FF' ;
    this.ctx.fillRect(0, 0, 16*40.5 , 16*40.5 );

    for( var y in mapa ){
      for( var x in mapa[y] ){
        var obj = mapa[x][y] ;
        switch( obj.type ){
          case 'wall'  : this.drawWall(x,y);     break;
          case 'apple' : this.drawApple(x,y) ;    break;
          case 'fly'   : this.drawFly(mapa,x,y); break;
        }
      }
    }

    $('#fly-info-container').html(
      'fliesToDo  : ' + this.showPoses( this.fliesToDo  ) + '<br>' + 
      'doneFlies  : ' + this.showPoses( this.doneFlies  ) + '<br>' + 
      'applePoses : ' + this.showPoses( this.applePoses ) + '<br>'  
      
      );

  },

  showPoses : function( poses ){
    
    var ret = '[ ';

    for( var i in poses ){
      ret += '(' + poses[i][0] + ',' + poses[i][1] + '), ' ;
    }

    return ret + ' ]' ;
  },

  drawFly : function( mapa , x_ , y_  ){

    var x = x_*16;
    var y = y_*16;

    this.ctx.drawImage( this.flyImg , x-8 , y );

    if( mapa[x_][y_].progName === '_'  ){
        this.ctx.beginPath();
        this.ctx.arc(x+10, y+10, 22 , 0, 2 * Math.PI, false);
        this.ctx.lineWidth = 1;
        this.ctx.strokeStyle = '#FFFF00';
        this.ctx.stroke();
    }

    this.ctx.fillStyle = "black";
    this.ctx.font = "bold 9px Arial";
    this.ctx.fillText( mapa[x_][y_].energy , x+25, y-10 );

  },

  drawApple : function(x,y){
    this.ctx.drawImage( this.appleImg , x*16 - 2 , y*16 -4);
  },

  drawWall : function(x,y){
    this.ctx.drawImage( this.wallImg , x*16 + 3  , y*16 +3 );
  },



};






});

File: v3/Problems/Fly02/Fly.hs
module Problems.Fly02.Fly where


import qualified Data.Map as Map
import Data.Map (Map)
import Data.List
import System.Random


import Problems.Fly02.Funs 
--(Energy,Pos,minus,dist,head')

data Input  = Input {
  myApplePoses    :: [Pos] ,
  nearestFlyPos   :: Maybe Pos ,
  myPos           :: Pos ,
  inputEnergy     :: Energy 
 } 

data Output = Output {
  moveDir :: Dir
} 


data Dir = DUp   
         | DDown 
         | DLeft 
         | DRight
         | DStay
  deriving ( Show )


prog_1 input = output_ $ posToDir_ (myPos_ input) (head_ $ myApplePoses_ input)
prog_2 _     = output_ $ dRight 
prog_3 input = output_ $ posToDir_ (myPos_ input) (nearestFlyPos_ input)

prog1 input = Output $ posToDir (myPos input) (head_ $ myApplePoses input)
prog2 _     = Output $ DRight
prog3 input = Output $ posToDir (myPos input) (nearestFlyPos input)
prog4 input = Output $ posToDir (myPos input) (avg $ myApplePoses input)



prog1' = ( prog1 , "prog1" )
prog2' = ( prog2 , "prog2" )
prog3' = ( prog3 , "prog3" )
prog4' = ( prog4 , "prog4" )





energies :: World -> [Energy]
energies w = map f (allFlyPoses w)
  where f pos = case objOnPos w pos of 
                 Fly flyData -> (flyEnergy flyData)
                 _ -> error $ "There shoud be a fly on " ++ show pos ++ "."



data World = World { 
  mapa       :: Map Pos Object ,
  fliesToDo  :: [Pos] ,
  doneFlies  :: [Pos] ,
  applePoses :: [Pos]
 }


data Object = Fly FlyData
            | Apple
            | Wall
            | Free


data FlyData = FlyData { 
  flyProg     :: Prog   ,
  flyProgName :: String ,
  flyEnergy   :: Energy 
 }

type Prog   = Input -> Output








steps :: Int -> World -> World
steps numSteps w = case numSteps of
 0 -> w
 n -> let w' = step w in steps (n-1) w' 

stepsList :: Int -> World -> [World]
stepsList numSteps w = case numSteps of
  0 -> [w]
  n -> let w' = step w in w : stepsList (n-1) w'

step :: World -> World
step w = case fliesToDo w of
  [] -> w{ fliesToDo = reverse (doneFlies w) ,
           doneFlies = [] }
  _  -> step $ stepCurrentFly w

stepCurrentFly :: World -> World
stepCurrentFly w = case (fliesToDo w) of
 flyPos:restFlies -> 
  case objOnPos w flyPos of
   Fly flyData -> 
     let input        = prepareInput w flyPos flyData
         output       = (flyProg flyData) input
         (w',flyPos') = tryToMoveFly (moveDir output) flyPos w
      in w'{ fliesToDo = restFlies ,
             doneFlies = flyPos' : (doneFlies w') }   
   _ -> error "There should be a fly!"
 [] -> error "No current fly!"
 
prepareInput :: World -> Pos -> FlyData -> Input
prepareInput w flyPos flyData = 
  Input { myApplePoses    = getSortedPoses w flyPos (applePoses w) ,
          nearestFlyPos   = getNearestFlyPos   w flyPos ,
          inputEnergy     = flyEnergy flyData ,
          myPos           = flyPos } 



getSortedPoses :: World -> Pos -> [Pos] -> [Pos]
getSortedPoses w pos poses = 
  map snd . sort . map (\pos'->(dist pos pos',pos')) $ poses

getNearestPos :: World -> Pos -> [Pos] -> Maybe Pos
getNearestPos w pos poses = case poses of
  [] -> Nothing
  _  -> Just . snd . minimum $ map (\pos'->(dist pos pos',pos')) $ poses

getNearestApplePos :: World -> Pos -> Maybe Pos
getNearestApplePos w pos = getNearestPos w pos (applePoses w)

getNearestFlyPos   :: World -> Pos -> Maybe Pos
getNearestFlyPos w pos = 
  let flies = (( fliesToDo w ++ doneFlies w ) \\  [pos] )
   in getNearestPos w pos flies 


tryToMoveFly :: Dir -> Pos -> World -> (World,Pos)
tryToMoveFly dir pos w = 
 let pos' = posPlusDir pos dir 
  in case objOnPos w pos' of
      Free -> ( moveFromTo pos pos' w , pos' )
      Apple-> ( eatApple   pos pos' w , pos' )
      _    -> ( w                     , pos  )


eatApple :: Pos -> Pos -> World -> World
eatApple flyPos applePos w0 = 
  let w1 = moveFromTo flyPos applePos w0
      w2 = updateEnergy (+1) applePos w1
   in w2 { applePoses = delete applePos (applePoses w2) }


updateEnergy :: (Energy->Energy) -> Pos -> World -> World
updateEnergy updateFun flyPos w = updateObj f flyPos w
 where f (Fly flyData) = Fly flyData{ flyEnergy = updateFun (flyEnergy flyData) } 
       f _             = error "Only fly has energy to update!"  

updateObj :: (Object->Object) -> Pos -> World -> World
updateObj updateFun pos w = 
  w { mapa = Map.update (Just . updateFun) pos (mapa w) }

moveFromTo :: Pos -> Pos -> World -> World
moveFromTo pos1 pos2 w = 
  let obj = objOnPos w pos1
   in putObjOnPos obj pos2 (deleteOnPos pos1 w)

posPlusDir :: Pos -> Dir -> Pos
posPlusDir (x,y) dir = case dir of
  DUp    -> (x  ,y-1)
  DDown  -> (x  ,y+1)
  DLeft  -> (x-1,y  )
  DRight -> (x+1,y  )
  DStay  -> (x  ,y  )

posToDir :: Pos -> Maybe Pos -> Dir
posToDir posMy Nothing = DStay 
posToDir posMy (Just posHer) 
  |   dx  > dy && (-dx) > dy = DUp
  |   dx  > dy               = DRight
  | (-dx) > dy               = DLeft
  | otherwise                = DDown
 where (dx,dy) = posHer `minus` posMy




w2 = foldr (uncurry putFly) w1 
 [ ( (10,10) , prog1' ) ,
   ( (30,30) , prog1' ) ]

putFly :: Pos -> (Prog , String) -> World -> World
putFly pos (prog,progName) w = 
  let flyData = FlyData { flyProg = prog , flyEnergy = 1 , flyProgName = progName }
      w'      = putObjOnPos (Fly flyData) pos w
   in w'{ fliesToDo = pos:(fliesToDo w') }


putApple :: Pos -> World -> World
putApple pos w = (putObjOnPos Apple pos w){
   applePoses = pos:(applePoses w)
 }



emptyWorld :: World
emptyWorld = World{
  mapa      = Map.empty ,
  fliesToDo = [] ,
  doneFlies = [] ,
  applePoses= [] 
 }

boxedWorld :: World
boxedWorld = foldr ($) emptyWorld 
 [ h (x,y) , 
   h (x,y+len-1) ,
   v (x,y+1) ,
   v (x+len-1,y+1)  ]
  where 
    (x,y) = (1,1)
    len = 38
    h = putHLine Wall len
    v = putVLine Wall (len-2)

w1 :: World  
w1 = foldr putApple boxedWorld applePoses
 where
  numApples = 20
  (xs,ys) = splitAt numApples $ take (2*numApples) $ randomRs (2,35) (mkStdGen 424242)
  applePoses = zip xs ys

objOnPos :: World -> Pos -> Object
objOnPos w pos = case Map.lookup pos (mapa w) of
  Nothing -> Free
  Just o  -> o

putObjOnPos :: Object -> Pos -> World -> World
putObjOnPos o pos w = w{ mapa = Map.insert pos o (mapa w) }

deleteOnPos :: Pos -> World -> World
deleteOnPos pos w = w{ mapa = Map.delete pos (mapa w) }

putHLine :: Object -> Int -> Pos -> World -> World
putHLine o len (x,y) w = foldr (putObjOnPos o) w points
  where points = [ (x',y) | x' <- [x..x+len-1] ]

putVLine :: Object -> Int -> Pos -> World -> World
putVLine o len (x,y) w = foldr (putObjOnPos o) w points
  where points = [ (x,y') | y' <- [y..y+len-1] ]


objChar :: Object -> Char
objChar o = case o of
 Fly _   -> 'F'
 Apple   -> 'A'
 Wall    -> 'W'
 Free    -> '.'  

instance Show World where
  show = showWorld defaultView

instance Show FlyData where
  show fd = "Energy : " ++ show (flyEnergy fd)

showWorld :: (Pos,Pos) -> World -> String
showWorld view@((x1,y1),(x2,y2)) w = 
   "\n" ++ 
   fliesInfo w ++ "\n" ++
   mapka ++ "\n\n" ++ 
   "fliesToDo  : " ++ show (fliesToDo  w) ++ "\n" ++ 
   "doneFlies  : " ++ show (doneFlies  w) ++ "\n" ++
   "applePoses : " ++ show (applePoses w) ++ "\n" 
  where
    mapka = intercalate "\n" $ map (concatMap (\o-> objChar o : " ")) $ worldToLists view w 



worldToLists :: (Pos,Pos) -> World -> [[Object]]
worldToLists ((x1,y1),(x2,y2)) w = 
  [ [ objOnPos w (x,y) | x <- [x1..x2] ] | y <- [y1..y2] ]


worldFromStrs :: [String] -> World
worldFromStrs strs = 
  foldr (\(pos,obj) w -> f obj pos w ) emptyWorld (worldFromStrs' strs)
 where
  f obj = case obj of
    Apple -> putApple 
    _     -> putObjOnPos obj

worldFromStrs' :: [String] -> [(Pos,Object)]
worldFromStrs' strs = 
  concatMap (\(str,y)-> map (\(ch,x)-> ((x,y),objectFromChar ch) ) (zip str [0..]) ) (zip strs [0..]) 

objectFromChar :: Char -> Object
objectFromChar ch = case ch of
  '.' -> Free
  'W' -> Wall
  'A' -> Apple



allFlyPoses :: World -> [Pos]
allFlyPoses w = reverse (doneFlies w) ++ (fliesToDo w)

fliesInfo :: World -> String
fliesInfo w = concatMap f (allFlyPoses w)
  where f pos = case objOnPos w pos of 
                 Fly flyData -> show pos ++ " : \t" ++ show flyData ++ "\n"
                 _ -> error $ "There shoud be a fly on " ++ show pos ++ "."





instance Show Object where
  show o = [ objChar o ]

narezej :: Int -> [a] -> [[a]]
narezej _ [] = []
narezej n xs = 
  let (ys,zs) = splitAt n xs 
   in ys : narezej n zs
 
defaultView :: (Pos,Pos)
defaultView = ((0,0),(39,39))


testShow = showWorld defaultView emptyWorld